### ISSUE TITLE
Multinomial classification metrics

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#664]
Existing classification metrics assume the true and predicted output are each a single class label. We'd like to extend them to multinomial predictions where for each data point, we are bucketing a fixed number of trials into different categories.

**Comment by cavaunpeu** [comment]
Hey @dustinvtran. I'd like to take this one. A few initial questions:

1. `if isinstance(output_key, binary_discrete + categorical_discrete)`, why would we ever use `n_samples > 1` to [generate `probs` in evaluate.py](https://github.com/blei-lab/edward/blob/91f63e49025006d0f5673f86835443fec6f89fc1/edward/criticisms/evaluate.py#L121)?

For multi-label data:

2. How should we [expand the `output_key` into `y_pred`](https://github.com/blei-lab/edward/blob/91f63e49025006d0f5673f86835443fec6f89fc1/edward/criticisms/evaluate.py#L128)? For instance, assume `output_key = Multinomial(total_count=5.0, probs=tf.constant([0.48, 0.51, 0.01]))`; what should `y_pred` be? What if `output_key = Multinomial(total_count=5.0, probs=tf.constant([0.33, 0.33, 0.34]))`?

3. Seemingly, the vanilla way to compute `categorical_accuracy` is to evaluate equivalence between `y_true` and `y_pred` along their outermost axes for each row (then return the mean). This is [how scikit-learn solves this problem](https://github.com/scikit-learn/scikit-learn/blob/d9fdd8b0d1053cb47af8e3823b7a05279dd72054/sklearn/metrics/classification.py#L171). Additionally, do we want to compute binary classification metrics *for each class* along this axis, then aggregate? Should the same aggregation strategies used in sckit-learn — dictated by the [`average` parameter in `_average_binary_score`](https://github.com/scikit-learn/scikit-learn/blob/d9fdd8b0d1053cb47af8e3823b7a05279dd72054/sklearn/metrics/base.py#L23) — be used?

4. Does it make sense to allow users to compute:
    - The mean squared error (and mean absolute error, etc.) between `y_true` and `y_pred`?
    - The categorical cross-entropy between something like `y_true / y_true.sum()` and [`probs`](https://github.com/blei-lab/edward/blob/91f63e49025006d0f5673f86835443fec6f89fc1/edward/criticisms/evaluate.py#L122)?

**Comment by dustinvtran** [comment]
1. Imagine `probs` is given by a prior or posterior distribution. Fetching `probs` returns a random value at each iteration. In general, we make predictions by Monte Carlo samples over the posterior predictive density.
2. I'm not sure how others do this; there's likely a standard. Here's one naive suggestion following this logic: if `total_count=5.0`, assign one count to a class every time the class has at least 0.20 probability; remove 0.20 probability from that class and repeat until all classes have < 0.20 probability. If there's one count remaining, choose the class with the highest remaining probability; else renormalize the probabilities and draw however remaining counts. This generalizes(?) the Bernoulli case when `probs=[0.5, 0.5]`.
3. I think including sklearn aggregation metrics makes sense. 
4. Yes for both.

**Comment by cavaunpeu** [comment]
@dustinvtran 

Apologies for late response. Wasn't "watching" repo so didn't receive notification.

1. The `probs` I'm referring to — and the only one that makes use of `n_samples` — is [here](https://github.com/blei-lab/edward/blob/91f63e49025006d0f5673f86835443fec6f89fc1/edward/criticisms/evaluate.py#L121) (and only here?). Therein, it seems that `output_key` is some distribution parameterized by a vector of probabilities π (e.g. a `Binomial` GLM). As such, `output_key.probs` returns π; when `n_samples > 1`, we simply get `[π] * n_samples`.

Does this seem right? Am I missing something?

2. Sounds great. I like your approach. I'll research this further as well. Finally, I'll implement this as a function/method so it's easy to change.

3. 👍.

4. 👍. Also, I'm guessing "we could also just use metrics that are only a function of `y_true` and `probs, total_count`" is equivalent to "(the categorical cross-entropy) between something like `y_true / y_true.sum()` and `probs`", as `y_true.sum()` should always equal `total_count`.

**Comment by dustinvtran** [comment]
> The probs I'm referring to — and the only one that makes use of n_samples — is here (and only here?). Therein, it seems that output_key is some distribution parameterized by a vector of probabilities π (e.g. a Binomial GLM). As such, output_key.probs returns π; when n_samples > 1, we simply get [π] * n_samples.

The line above it says that `output_key` is a discrete random variable. It's parameterized by `probs`, which can itself be a random variable. Therefore as I mentioned, we need to fetch it many times and average.

**Comment by cavaunpeu** [comment]
> "It's parameterized by `probs`, which can itself be a random variable."

Ah, ok. This makes sense. Thanks for the clarification!
